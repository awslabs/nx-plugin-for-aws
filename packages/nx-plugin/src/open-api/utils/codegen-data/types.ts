/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { Plugin } from '@hey-api/openapi-ts';
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type ClientData = Parameters<Plugin.LegacyHandler<any>>[0]['client'];
export type Operation = ClientData['services'][number]['operations'][number];
export type Model = ClientData['models'][number];

export interface CodeGenData extends ClientData {
  className: string;
  info: OpenAPIV3.InfoObject | OpenAPIV3_1.InfoObject;
  allOperations: Operation[];
  operationsByTag: { [tag: string]: Operation[] };
  untaggedOperations: Operation[];
  vendorExtensions: { [key: string]: any };
}

export const flattenModelLink = (link?: Model | Model[]): Model =>
  link === undefined ? undefined : Array.isArray(link) ? link[0] : link;

// Model types which indicate it is composed (ie inherits/mixin's another schema)
export const COMPOSED_SCHEMA_TYPES = new Set(['one-of', 'any-of', 'all-of']);
export const COLLECTION_TYPES = new Set(['array', 'dictionary']);
export const PRIMITIVE_TYPES = new Set([
  'string',
  'integer',
  'number',
  'boolean',
  'null',
  'any',
  'binary',
  'void',
]);
