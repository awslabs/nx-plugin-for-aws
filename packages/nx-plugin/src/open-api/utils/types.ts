/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type Spec = OpenAPIV3.Document | OpenAPIV3_1.Document;
