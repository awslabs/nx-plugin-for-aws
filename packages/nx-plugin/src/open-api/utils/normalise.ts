/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import cloneDeepWith from 'lodash.clonedeepwith';
import type { OpenAPIV3 } from 'openapi-types';
import { isRef, resolveIfRef, resolveRef, splitRef } from './refs';
import { Spec } from './types';
import { pascalCase, upperFirst } from '../../utils/names';
import camelCase from 'lodash.camelcase';

interface SubSchema {
  readonly nameParts: string[];
  readonly schema: OpenAPIV3.SchemaObject;
  readonly propPath: (string | number)[];
}

interface SubSchemaRef {
  readonly $ref: string;
  readonly name: string;
  readonly schema: OpenAPIV3.SchemaObject;
}

const isCompositeSchema = (schema: OpenAPIV3.SchemaObject) =>
  !!schema.allOf || !!schema.anyOf || !!schema.oneOf;

const hasSubSchemasToVisit = (
  schema?: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject,
): schema is OpenAPIV3.SchemaObject =>
  !!schema &&
  !isRef(schema) &&
  (['object', 'array'].includes(schema.type as any) ||
    isCompositeSchema(schema) ||
    !!schema.not ||
    (schema.type === 'string' && !!schema.enum));

const filterInlineCompositeSchemas = (
  schemas: (OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject)[],
  nameParts: string[],
  namePartPrefix: string,
  propPath: (string | number)[],
): SubSchema[] => {
  let inlineSchemaIndex = 0;
  return schemas.flatMap((s, i) => {
    if (hasSubSchemasToVisit(s)) {
      const subSchema: SubSchema = {
        nameParts: s.title
          ? [pascalCase(s.title)]
          : [
              ...nameParts,
              `${namePartPrefix}${inlineSchemaIndex === 0 ? '' : inlineSchemaIndex}`,
            ],
        schema: s,
        propPath: [...propPath, i],
      };
      inlineSchemaIndex++;
      return [subSchema];
    }
    return [];
  });
};

const hoistInlineObjectSubSchemas = (
  nameParts: string[],
  schema: OpenAPIV3.SchemaObject,
): SubSchemaRef[] => {
  // Find all the inline subschemas we should visit
  const inlineSubSchemas: SubSchema[] = [
    ...(hasSubSchemasToVisit(schema.not)
      ? [
          {
            nameParts: schema.not?.title
              ? [pascalCase(schema.not?.title)]
              : [...nameParts, 'Not'],
            schema: schema.not,
            propPath: ['not'],
          },
        ]
      : []),
    ...(schema.anyOf
      ? filterInlineCompositeSchemas(schema.anyOf, nameParts, 'AnyOf', [
          'anyOf',
        ])
      : []),
    ...(schema.allOf
      ? filterInlineCompositeSchemas(schema.allOf, nameParts, 'AllOf', [
          'allOf',
        ])
      : []),
    ...(schema.oneOf
      ? filterInlineCompositeSchemas(schema.oneOf, nameParts, 'OneOf', [
          'oneOf',
        ])
      : []),
    ...('items' in schema && hasSubSchemasToVisit(schema.items)
      ? [
          {
            nameParts: schema.items?.title
              ? [pascalCase(schema.items?.title)]
              : [...nameParts, 'Item'],
            schema: schema.items,
            propPath: ['items'],
          },
        ]
      : []),
    ...Object.entries(schema.properties ?? {})
      .filter(([, s]) => hasSubSchemasToVisit(s))
      .map(([name, s]) => ({
        nameParts: (s as OpenAPIV3.SchemaObject).title
          ? [pascalCase((s as OpenAPIV3.SchemaObject).title)]
          : [...nameParts, name],
        schema: s as OpenAPIV3.SchemaObject,
        propPath: ['properties', name],
      })),
    ...(typeof schema.additionalProperties !== 'boolean' &&
    hasSubSchemasToVisit(schema.additionalProperties)
      ? [
          {
            nameParts: schema.additionalProperties?.title
              ? [pascalCase(schema.additionalProperties?.title)]
              : [...nameParts, 'Value'],
            schema: schema.additionalProperties,
            propPath: ['additionalProperties'],
          },
        ]
      : []),
  ];

  // Hoist these recursively first (ie depth first search) so that we don't miss refs
  const recursiveRefs = inlineSubSchemas.flatMap((s) =>
    hoistInlineObjectSubSchemas(s.nameParts, s.schema),
  );

  // Clone the object subschemas to build the refs. Note that only objects with "properties" are hoisted as these are non-dictionary types
  const refs = inlineSubSchemas
    .filter(
      (s) =>
        (s.schema.type === 'object' && s.schema.properties) ||
        isCompositeSchema(s.schema) ||
        (s.schema.type === 'string' && s.schema.enum),
    )
    .map((s) => {
      const name = s.nameParts.map(upperFirst).join('');
      const $ref = `#/components/schemas/${name}`;
      const ref = {
        $ref,
        name,
        schema: structuredClone({
          ...s.schema,
          'x-aws-nx-hoisted': true,
        }),
      };

      // Replace each subschema with a ref in the spec
      const schemaWithPropToReplace = s.propPath
        .slice(0, -1)
        .reduce(
          (resolvedInSchema, pathPart) => resolvedInSchema?.[pathPart],
          schema,
        );
      if (schemaWithPropToReplace) {
        schemaWithPropToReplace[s.propPath[s.propPath.length - 1]] = { $ref };
      }

      return ref;
    });

  return [...refs, ...recursiveRefs];
};

/**
 * In order to ensure we generate models consistently whether or not users used refs or inline schemas,
 * we hoist any inline refs to non-primitives
 */
export const normaliseOpenApiSpecForCodeGen = (inSpec: Spec): Spec => {
  // Clone the spec so we're free to mutate it
  let spec = cloneDeepWith(inSpec);

  // Ensure spec has schemas set
  if (!spec?.components?.schemas) {
    spec.components = {
      ...spec.components,
    };
    spec.components.schemas = {
      ...spec.components.schemas,
    };
  }

  const seenOperationIds = new Set<string>();
  const duplicatedOperationIds = new Set<string>();

  // Make sure all operationIds are camelCase, and find which ones are duplicated
  Object.entries(spec.paths ?? {}).forEach(([path, pathOps]) =>
    Object.entries(pathOps ?? {}).forEach(([method, op]) => {
      const operation = resolveIfRef(spec, op);
      if (operation && typeof operation === 'object') {
        const operationId = camelCase(
          (operation as any).operationId ?? `${method}-${path}`,
        );
        (operation as any).operationId = operationId;
        if (seenOperationIds.has(operationId)) {
          duplicatedOperationIds.add(operationId);
        }
        seenOperationIds.add(operationId);
      }
    }),
  );

  // Reset seen operation ids
  seenOperationIds.clear();

  const untagged = Symbol('untagged');
  const seenOperationIdsByTag: { [tag: string | symbol]: Set<string> } = {};

  // "Hoist" inline request and response schemas
  Object.entries(spec.paths ?? {}).forEach(([path, pathOps]) =>
    Object.entries(pathOps ?? {}).forEach(([method, op]) => {
      const operation = resolveIfRef(spec, op);
      if (operation && typeof operation === 'object') {
        const tags: string[] = (operation as any).tags ?? [];
        const operationId = (operation as any).operationId as string;

        // Allow operations to have the same id
        let deduplicatedOpId = operationId;

        // Attempt to deduplicate the operationId by its tags
        if (duplicatedOperationIds.has(operationId)) {
          deduplicatedOpId = camelCase(
            tags.map((t) => `${t}-`).join('') + operationId,
          );
        }

        (operation as any)['x-aws-nx-deduplicated-op-id'] = deduplicatedOpId;

        seenOperationIds.add(deduplicatedOpId);

        // Throw an error for any duplicated operation ids with the same tag, or untagged
        [...(tags.length === 0 ? [untagged] : tags)].forEach((tag) => {
          if (
            seenOperationIdsByTag[tag] &&
            seenOperationIdsByTag[tag].has(operationId)
          ) {
            throw new Error(
              tag === untagged
                ? `Untagged operations cannot have the same operationId (${operationId})`
                : `Operations with the same tag (${String(tag)}) cannot have the same operationId (${operationId})`,
            );
          }

          seenOperationIdsByTag[tag] = new Set([
            ...(seenOperationIdsByTag[tag] ?? []),
            operationId,
          ]);
        });

        if ('responses' in operation) {
          Object.entries(operation.responses ?? {}).forEach(([code, res]) => {
            const response = resolveIfRef(spec, res);
            const jsonResponseSchema =
              response?.content?.['application/json']?.schema;
            if (
              jsonResponseSchema &&
              !isRef(jsonResponseSchema) &&
              (['object', 'array'].includes(jsonResponseSchema.type!) ||
                isCompositeSchema(jsonResponseSchema) ||
                (jsonResponseSchema?.type === 'string' &&
                  jsonResponseSchema.enum))
            ) {
              const schemaName = `${upperFirst(deduplicatedOpId)}${code}Response`;
              spec.components!.schemas![schemaName] = jsonResponseSchema;
              response!.content!['application/json'].schema = {
                $ref: `#/components/schemas/${schemaName}`,
              };
            }
          });
        }
        if ('requestBody' in operation) {
          const requestBody = resolveIfRef(spec, operation.requestBody);
          const jsonRequestSchema =
            requestBody?.content?.['application/json']?.schema;
          if (
            jsonRequestSchema &&
            !isRef(jsonRequestSchema) &&
            (['object', 'array'].includes(jsonRequestSchema.type!) ||
              isCompositeSchema(jsonRequestSchema) ||
              (jsonRequestSchema?.type === 'string' && jsonRequestSchema.enum))
          ) {
            const schemaName = `${upperFirst(deduplicatedOpId)}RequestContent`;
            spec.components!.schemas![schemaName] = jsonRequestSchema;
            requestBody!.content!['application/json'].schema = {
              $ref: `#/components/schemas/${schemaName}`,
            };
          }
        }
      }
    }),
  );

  // "Hoist" any nested object definitions in arrays/maps that aren't already refs, as parseOpenapi will treat the
  // type as "any" if they're defined inline (and not a ref)
  Object.entries(spec.components?.schemas ?? {}).forEach(([name, schema]) => {
    if (!isRef(schema)) {
      const refs = hoistInlineObjectSubSchemas([name], schema);
      refs.forEach((ref) => {
        spec.components!.schemas![ref.name] = ref.schema;
      });
    }
  });

  // "Inline" any refs to non objects/enums
  const inlinedRefs: Set<string> = new Set();
  spec = cloneDeepWith(spec, (v) => {
    if (v && typeof v === 'object' && v.$ref) {
      const resolved = resolveRef(spec, v.$ref);
      if (
        resolved &&
        resolved.type &&
        resolved.type !== 'object' &&
        !(resolved.type === 'string' && resolved.enum)
      ) {
        inlinedRefs.add(v.$ref);
        return resolved;
      }
    }
  });

  // Delete the non object schemas that were inlined
  [...inlinedRefs].forEach((ref) => {
    const parts = splitRef(ref);
    if (
      parts.length === 3 &&
      parts[0] === 'components' &&
      parts[1] === 'schemas'
    ) {
      delete spec.components!.schemas![parts[2]];
    }
  });

  return spec;
};
