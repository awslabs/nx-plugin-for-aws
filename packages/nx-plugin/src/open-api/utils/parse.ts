/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import SwaggerParser from '@apidevtools/swagger-parser';
import { Spec } from './types';

const TREE_PROTOCOL_PREFIX = 'workspace://';

/**
 * Parse an OpenAPI specification given its path within the tree
 */
export const parseOpenApiSpec = async (
  tree: Tree,
  specPath: string,
): Promise<Spec> => {
  // Use a custom protocol prefix to ensure that the swagger parser doesn't try and read from the filesystem itself
  const spec = await SwaggerParser.bundle(
    `${TREE_PROTOCOL_PREFIX}/${specPath}`,
    {
      resolve: {
        // Use a custom resolver which reads files from the tree rather than directly from the filesystem
        file: {
          order: 1,
          read: (file) => {
            const path = file.url.slice(TREE_PROTOCOL_PREFIX.length);
            if (!tree.exists(path)) {
              const message = `Unable to find ${specPath} in workspace ${tree.root}`;

              // Write an error to console as swagger parser's error message may be confusing with our custom protocol prefix
              console.error(message);
              throw new Error(message);
            }
            return tree.read(path);
          },
          canRead: (file) => file.url.startsWith(TREE_PROTOCOL_PREFIX),
        },
        http: false,
      },
    },
  );

  if ('swagger' in spec) {
    throw new Error(
      `OpenAPI v2 specifications are not supported. Please use OpenAPI v3`,
    );
  }
  return spec;
};
