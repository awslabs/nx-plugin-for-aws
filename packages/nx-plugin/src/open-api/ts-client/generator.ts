/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiTsClientGeneratorSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import * as path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';
import { formatFilesInSubtree } from '../../utils/format';

/**
 * Generates typescript client from an openapi spec
 */
export const openApiTsClientGenerator = async (
  tree: Tree,
  options: OpenApiTsClientGeneratorSchema,
) => {
  const data = await buildOpenApiCodeGenerationData(
    tree,
    options.openApiSpecPath,
  );

  generateOpenApiTsClient(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Build a data structure which can be used to generate code from OpenAPI
 */
export const buildOpenApiCodeGenerationData = async (
  tree: Tree,
  specPath: string,
) => {
  const spec = await parseOpenApiSpec(tree, specPath);
  return await buildOpenApiCodeGenData(spec);
};

/**
 * Generate an OpenAPI typescript client in the target directory
 */
export const generateOpenApiTsClient = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsClientGenerator;
