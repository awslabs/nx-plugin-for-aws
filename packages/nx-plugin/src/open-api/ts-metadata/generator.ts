/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiMetadataSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { formatFilesInSubtree } from '../../utils/format';
import path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';

/**
 * Generates typescript metadata from an openapi spec
 */
export const openApiTsMetadataGenerator = async (
  tree: Tree,
  options: OpenApiMetadataSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsMetadata(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Generate OpenAPI typescript metadata in the target directory
 */
export const generateOpenApiTsMetadata = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsMetadataGenerator;
