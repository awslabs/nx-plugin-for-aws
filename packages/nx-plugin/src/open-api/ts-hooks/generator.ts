/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { OpenApiHooksSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { generateOpenApiTsClient } from '../ts-client/generator';

/**
 * Generates typescript hooks from an openapi spec
 */
export const openApiTsHooksGenerator = async (
  tree: Tree,
  options: OpenApiHooksSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsClient(tree, data, options.outputPath);

  // TODO: generate hooks which wrap the client
};

export default openApiTsHooksGenerator;
