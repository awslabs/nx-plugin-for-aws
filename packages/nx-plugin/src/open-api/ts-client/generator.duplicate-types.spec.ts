/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { openApiTsClientGenerator } from './generator';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { Spec } from '../utils/types';
import { Tree } from '@nx/devkit';
import { expectTypeScriptToCompile } from '../../utils/test/ts.spec';

describe('openApiTsClientGenerator - duplicate types', () => {
  let tree: Tree;
  const title = 'TestApi';

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  const validateTypeScript = (paths: string[]) => {
    expectTypeScriptToCompile(tree, paths);
  };

  it('should handle operation ID matching schema name conflict', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      components: {
        schemas: {
          ChatRequest: {
            type: 'object',
            properties: {
              message: { type: 'string' },
            },
            required: ['message'],
          },
        },
      },
      paths: {
        '/chat': {
          post: {
            operationId: 'chat',
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ChatRequest',
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'Success',
                content: {
                  'application/json': {
                    schema: { type: 'string' },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsClientGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
    ]);

    const types = tree.read('src/generated/types.gen.ts', 'utf-8');
    expect(types).toMatchSnapshot();

    const client = tree.read('src/generated/client.gen.ts', 'utf-8');
    expect(client).toMatchSnapshot();

    // Verify no duplicate type declarations
    const typeMatches = types?.match(/export type ChatRequest/g) || [];
    expect(typeMatches).toHaveLength(1);
  });

  it('should handle multiple operation ID conflicts', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      components: {
        schemas: {
          UserRequest: {
            type: 'object',
            properties: {
              name: { type: 'string' },
            },
            required: ['name'],
          },
          OrderRequest: {
            type: 'object',
            properties: {
              item: { type: 'string' },
            },
            required: ['item'],
          },
        },
      },
      paths: {
        '/user': {
          post: {
            operationId: 'user',
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/UserRequest',
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'Success',
                content: {
                  'application/json': {
                    schema: { type: 'string' },
                  },
                },
              },
            },
          },
        },
        '/order': {
          post: {
            operationId: 'order',
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/OrderRequest',
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'Success',
                content: {
                  'application/json': {
                    schema: { type: 'string' },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsClientGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
    ]);

    const types = tree.read('src/generated/types.gen.ts', 'utf-8');
    expect(types).toMatchSnapshot();

    const client = tree.read('src/generated/client.gen.ts', 'utf-8');
    expect(client).toMatchSnapshot();

    // Verify no duplicate type declarations
    const userTypeMatches = types?.match(/export type UserRequest/g) || [];
    expect(userTypeMatches).toHaveLength(1);

    const orderTypeMatches = types?.match(/export type OrderRequest/g) || [];
    expect(orderTypeMatches).toHaveLength(1);
  });
});
