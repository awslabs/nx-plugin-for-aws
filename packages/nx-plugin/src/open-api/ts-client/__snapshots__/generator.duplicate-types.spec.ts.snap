// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openApiTsClientGenerator - duplicate types > should handle multiple operation ID conflicts 1`] = `
"export type OrderRequest = {
  item: string;
};
export type UserRequest = {
  name: string;
};

export type OrderOperationRequest = OrderRequest;
export type OrderError = never;

export type UserOperationRequest = UserRequest;
export type UserError = never;
"
`;

exports[`openApiTsClientGenerator - duplicate types > should handle multiple operation ID conflicts 2`] = `
"import type {
  OrderRequest,
  UserRequest,
  OrderOperationRequest,
  UserOperationRequest,
} from './types.gen.js';

/**
 * Utility for serialisation and deserialisation of API types.
 */
export class $IO {
  protected static $mapValues = (data: any, fn: (item: any) => any) =>
    Object.fromEntries(Object.entries(data).map(([k, v]) => [k, fn(v)]));

  public static OrderRequest = {
    toJson: (model: OrderRequest): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.item === undefined
          ? {}
          : {
              item: model.item,
            }),
      };
    },
    fromJson: (json: any): OrderRequest => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        item: json['item'],
      };
    },
  };

  public static UserRequest = {
    toJson: (model: UserRequest): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.name === undefined
          ? {}
          : {
              name: model.name,
            }),
      };
    },
    fromJson: (json: any): UserRequest => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        name: json['name'],
      };
    },
  };
}

/**
 * Client configuration for TestApi
 */
export interface TestApiConfig {
  /**
   * Base URL for the API
   */
  url: string;
  /**
   * Custom instance of fetch. By default the global 'fetch' is used.
   * You can override this to add custom middleware for use cases such as adding authentication headers.
   */
  fetch?: typeof fetch;
  /**
   * Additional configuration
   */
  options?: {
    /**
     * By default, the client will add a Content-Type header, set to the media type defined for
     * the request in the OpenAPI specification.
     * Set this to false to omit this header.
     */
    omitContentTypeHeader?: boolean;
  };
}

/**
 * API Client for TestApi
 */
export class TestApi {
  private $config: TestApiConfig;

  constructor(config: TestApiConfig) {
    this.$config = config;

    this.order = this.order.bind(this);
    this.user = this.user.bind(this);
  }

  private $url = (
    path: string,
    pathParameters: { [key: string]: any },
    queryParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): string => {
    const baseUrl = this.$config.url.endsWith('/')
      ? this.$config.url.slice(0, -1)
      : this.$config.url;
    const pathWithParameters = Object.entries(pathParameters).reduce(
      (withParams, [key, value]) =>
        withParams.replace(\`{\${key}}\`, encodeURIComponent(\`\${value}\`)),
      path,
    );
    const queryString = Object.entries(queryParameters)
      .map(([key, value]) => {
        if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
          return value
            .map(
              (v) => \`\${encodeURIComponent(key)}=\${encodeURIComponent(\`\${v}\`)}\`,
            )
            .join('&');
        }
        return \`\${encodeURIComponent(key)}=\${encodeURIComponent(Array.isArray(value) ? value.map(String).join(',') : String(value))}\`;
      })
      .join('&');
    return (
      baseUrl + pathWithParameters + (queryString ? \`?\${queryString}\` : '')
    );
  };

  private $headers = (
    headerParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): [string, string][] => {
    return Object.entries(headerParameters).flatMap(([key, value]) => {
      if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
        return value.map((v) => [key, String(v)]) as [string, string][];
      }
      return [[key, String(value)]];
    });
  };

  private $fetch: typeof fetch = (...args) =>
    (this.$config.fetch ?? fetch)(...args);

  public async order(input: OrderOperationRequest): Promise<string> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } = {};
    const headerParameters: { [key: string]: any } = {};
    if (!this.$config.options?.omitContentTypeHeader) {
      headerParameters['Content-Type'] = 'application/json';
    }
    const body =
      typeof input === 'object'
        ? JSON.stringify($IO.OrderRequest.toJson(input))
        : String($IO.OrderRequest.toJson(input));

    const response = await this.$fetch(
      this.$url('/order', pathParameters, queryParameters),
      {
        headers: this.$headers(headerParameters),
        method: 'POST',
        body,
      },
    );

    if (response.status === 200) {
      return await response.text();
    }
    throw new Error(
      \`Unknown response status \${response.status} returned by API\`,
    );
  }

  public async user(input: UserOperationRequest): Promise<string> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } = {};
    const headerParameters: { [key: string]: any } = {};
    if (!this.$config.options?.omitContentTypeHeader) {
      headerParameters['Content-Type'] = 'application/json';
    }
    const body =
      typeof input === 'object'
        ? JSON.stringify($IO.UserRequest.toJson(input))
        : String($IO.UserRequest.toJson(input));

    const response = await this.$fetch(
      this.$url('/user', pathParameters, queryParameters),
      {
        headers: this.$headers(headerParameters),
        method: 'POST',
        body,
      },
    );

    if (response.status === 200) {
      return await response.text();
    }
    throw new Error(
      \`Unknown response status \${response.status} returned by API\`,
    );
  }
}
"
`;

exports[`openApiTsClientGenerator - duplicate types > should handle operation ID matching schema name conflict 1`] = `
"export type ChatRequest = {
  message: string;
};

export type ChatOperationRequest = ChatRequest;
export type ChatError = never;
"
`;

exports[`openApiTsClientGenerator - duplicate types > should handle operation ID matching schema name conflict 2`] = `
"import type { ChatRequest, ChatOperationRequest } from './types.gen.js';

/**
 * Utility for serialisation and deserialisation of API types.
 */
export class $IO {
  protected static $mapValues = (data: any, fn: (item: any) => any) =>
    Object.fromEntries(Object.entries(data).map(([k, v]) => [k, fn(v)]));

  public static ChatRequest = {
    toJson: (model: ChatRequest): any => {
      if (model === undefined || model === null) {
        return model;
      }
      return {
        ...(model.message === undefined
          ? {}
          : {
              message: model.message,
            }),
      };
    },
    fromJson: (json: any): ChatRequest => {
      if (json === undefined || json === null) {
        return json;
      }
      return {
        message: json['message'],
      };
    },
  };
}

/**
 * Client configuration for TestApi
 */
export interface TestApiConfig {
  /**
   * Base URL for the API
   */
  url: string;
  /**
   * Custom instance of fetch. By default the global 'fetch' is used.
   * You can override this to add custom middleware for use cases such as adding authentication headers.
   */
  fetch?: typeof fetch;
  /**
   * Additional configuration
   */
  options?: {
    /**
     * By default, the client will add a Content-Type header, set to the media type defined for
     * the request in the OpenAPI specification.
     * Set this to false to omit this header.
     */
    omitContentTypeHeader?: boolean;
  };
}

/**
 * API Client for TestApi
 */
export class TestApi {
  private $config: TestApiConfig;

  constructor(config: TestApiConfig) {
    this.$config = config;

    this.chat = this.chat.bind(this);
  }

  private $url = (
    path: string,
    pathParameters: { [key: string]: any },
    queryParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): string => {
    const baseUrl = this.$config.url.endsWith('/')
      ? this.$config.url.slice(0, -1)
      : this.$config.url;
    const pathWithParameters = Object.entries(pathParameters).reduce(
      (withParams, [key, value]) =>
        withParams.replace(\`{\${key}}\`, encodeURIComponent(\`\${value}\`)),
      path,
    );
    const queryString = Object.entries(queryParameters)
      .map(([key, value]) => {
        if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
          return value
            .map(
              (v) => \`\${encodeURIComponent(key)}=\${encodeURIComponent(\`\${v}\`)}\`,
            )
            .join('&');
        }
        return \`\${encodeURIComponent(key)}=\${encodeURIComponent(Array.isArray(value) ? value.map(String).join(',') : String(value))}\`;
      })
      .join('&');
    return (
      baseUrl + pathWithParameters + (queryString ? \`?\${queryString}\` : '')
    );
  };

  private $headers = (
    headerParameters: { [key: string]: any },
    collectionFormats?: { [key: string]: 'multi' | 'csv' },
  ): [string, string][] => {
    return Object.entries(headerParameters).flatMap(([key, value]) => {
      if (Array.isArray(value) && collectionFormats?.[key] === 'multi') {
        return value.map((v) => [key, String(v)]) as [string, string][];
      }
      return [[key, String(value)]];
    });
  };

  private $fetch: typeof fetch = (...args) =>
    (this.$config.fetch ?? fetch)(...args);

  public async chat(input: ChatOperationRequest): Promise<string> {
    const pathParameters: { [key: string]: any } = {};
    const queryParameters: { [key: string]: any } = {};
    const headerParameters: { [key: string]: any } = {};
    if (!this.$config.options?.omitContentTypeHeader) {
      headerParameters['Content-Type'] = 'application/json';
    }
    const body =
      typeof input === 'object'
        ? JSON.stringify($IO.ChatRequest.toJson(input))
        : String($IO.ChatRequest.toJson(input));

    const response = await this.$fetch(
      this.$url('/chat', pathParameters, queryParameters),
      {
        headers: this.$headers(headerParameters),
        method: 'POST',
        body,
      },
    );

    if (response.status === 200) {
      return await response.text();
    }
    throw new Error(
      \`Unknown response status \${response.status} returned by API\`,
    );
  }
}
"
`;
